ao-tools/ao-sendfake/ao-sendfake.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-LOAD 1 "ao-sendfake" ""
.SH NAME
ao-sendfake \- flash a program to an ARM-based AltOS device
.SH SYNOPSIS
.B "ao-sendfake"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
[\--cal \fIradio-calibration\fP]
[\--serial \fserial-number\fP]
\fIfile.elf\fP or \fIfile.ihx\fP
.SH DESCRIPTION
.I ao-sendfake
loads the specified .elf or .ihx file into the target device flash
memory via the AltOS boot loader, using either existing serial number
and radio calibration values or taking either of those from the
command line.
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device the debugger uses to communicate with
the target device. The special name 'BITBANG' directs ao-dbg to use
the cp2103 connection, otherwise this should be a usb serial port
connected to a suitable cc1111 debug node.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMega:2
.br
TeleMega
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.TP
\-s serial-number | --serial serial-number
This programs the device serial number into the image. If no serial
number is specified, then the existing serial number, if any, will be
read from the device.
.TP
\-c radio-calibration | --cal radio-calibration This programs the
radio calibration value into the image for hardware which doesn't have
any eeprom storage for this value. If no calibration value is
specified, an existing calibration value will be used. The value here
can be computed given the current radio calibration value, the
measured frequency and the desired frequency:
.IP
	cal' = cal * (desired/measured)
.IP
The default calibration value is 7119667.
.SH USAGE
.I ao-sendfake
reads the specified .elf file into memory, edits the image to
customize it using the specified serial number and radio calibration
values. It then connects to the debug dongle and writes the program to
the target device flash memory.
.SH AUTHOR
Keith Packard


ChangeLog
commit 701c26ed85c28ac59e338975f2a6ba6bd25f6493
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 19 00:16:16 2013 -0800

    altosdroid: bump versionName to 1.3 and versionCode to 4
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9f95ffbad918a73cfd5460d6ce037d680465c35d
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 19 00:12:21 2013 -0800

    altosui: When device has no valid romconfig, set RF cal to 0
    
    This is intended to signal to the user that no valid value was found
    and that they'd best pick something sensible.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a04c1dd5df76c9127615bc797a9d9f764eec1234
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 19 00:08:50 2013 -0800

    altos/lpc: Stop sending SETUP IN when the requested size is reached
    
    The host won't keep asking for SETUP IN packets once it has received
    the amount of data requested, so check to see if we've sent that much
    and flip back to IDLE state if so.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1ab12861c3e70d7c22b27d988546a925616a0adc
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 23:27:34 2013 -0800

    altos/lpc: Reset less of the device on USB reset.
    
    This leaves most of the device configured across USB reset, which
    appears to help when sending a IN reply to the first SETUP packet;
    without this change, the IN reply would always get a length of 0,
    which is fine for SET_ADDRESS, but not for GET_DESCRIPTOR_DEVICE,
    which OS X appears to send before setting the address (go figure).
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 3b13cc2ca035b13582cd2e59ba7286f872f43c6e
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 22:00:13 2013 -0800

    altoslib: Remove some old debug printfs for self flashing
    
    These aren't necessary anymore and just slow down flashing boards.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 408b0dea338147382e94717dab85b4a204e7bdf5
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 21:08:33 2013 -0800

    micropeak: Add micropeak man page
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2b2ba87d5f68b9e052dddd49d69341f36d777122
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 21:02:15 2013 -0800

    ao-tools: Add man pages for ao-flash utilities
    
    These aren't very wordy, but these tools are pretty simple scripts.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 39cb8c2896317b7538353be979ac99baffc14489
Merge: 2a6016c ee42796
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 21:53:52 2013 -0700

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 2a6016cfabc8cd56f5219871e3b3df316a639289
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 21:53:36 2013 -0700

    update Debian standards version we claim compliance with

commit ee4279613b4757453d0d8f8afc06037c61eeb520
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 20:32:05 2013 -0800

    altos: Try IMU self-test 10 times before giving up
    
    This should keep the device from failing to boot unless the IMU is
    actually broken. Oh, and if self test does fail, this places the
    flight computer in 'Invalid' state rather than panic.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1bf84ec28a41f7bd1b11ba45b4639856266227bc
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 20:30:58 2013 -0800

    doc: Add tables describing AltOS beeps and flashes
    
    Provide a convenient place to reference when listening to the device.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0673344289772ed89483948184d6608c272c7c26
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 18:20:55 2013 -0800

    altos/stm: Semantic error in STM usb disable caused it to not work
    
    The USB enable register wasn't actually getting rewritten with the
    enable bit turned off, so the USB device was still powered on in flight.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 122f491e459b6ff417932370b3f1aa2091c71aca
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 18:30:54 2013 -0700

    update release docs to include option for submodules

commit d9982c257463f23be940eea66bd4dc3aadff0043
Merge: 1b97ed2 b63fc05
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 18:25:35 2013 -0700

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 1b97ed2b64bcbcd969124964f1e49837899f1c70
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 18 18:25:03 2013 -0700

    we're using packaged and local-to-our-tree ARM toolchain now

commit b63fc05481bf6d57e6385704ce53c1c19afa9c2e
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:34:31 2013 -0800

    doc: typo in micropeak doc hole->hold

commit 6827961c002757f8e74de44f6eb9c9029d099ebc
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:25:41 2013 -0800

    doc: Update micropeak quick start guide to note new boost detect
    
    Now waits for one minute and 30m of altitude change to avoid false detections.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c0966cd40f05f3a65b0c977b4b92586a58192f4b
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:22:51 2013 -0800

    micropeak: Compile for java 6
    
    Don't a require later version as not all target OSes support it
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit eea036650e62bc0f8652155974b512686754fd13
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 14:08:41 2013 -0800

    Move pdclib build results to pdclib-root
    
    This makes pdclib easier to manage as a submodule
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c1bfe09b6d3eb28d0c7cfe07a248843cf81bcd25
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Dec 18 13:36:04 2013 -0800


ao-tools/ao-sendfake/ao-sendfake.c
/*
 * Copyright © 2012 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <err.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sysexits.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <stdbool.h>
#include "ccdbg.h"
#include "cc-usb.h"
#include "cc.h"
#include "ao-verbose.h"


static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "verbose", .has_arg = 0, .val = 'v' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--device=<device>] [-tty=<tty>] [--verbose] file\n", program);
	exit(1);
}

void
done(struct cc_usb *cc, int code)
{
/*	cc_usb_printf(cc, "a\n"); */
	cc_usb_close(cc);
	exit (code);
}

int
main (int argc, char **argv)
{
	char			*device = NULL;
	char			*filename;
	unsigned int		s;
	int			i;
	int			string_num;
	int			c;
	int			tries;
	struct cc_usb		*cc = NULL;
	char			*tty = NULL;
	int			success;
	int			verbose = 0;
	FILE			*file;
	char			buf[1024];

	while ((c = getopt_long(argc, argv, "vT:D:", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}

	ao_verbose = verbose;

	if (verbose > 1)
		ccdbg_add_debug(CC_DEBUG_BITBANG);

	filename = argv[optind];
	if (filename == NULL)
		usage(argv[0]);

	file = fopen(filename, "r");
	if (!file) {
		perror (filename);
		usage(argv[0]);
	}

	if (!tty)
		tty = cc_usbdevs_find_by_arg(device, "TeleMega");
	if (!tty)
		tty = getenv("ALTOS_TTY");
	if (!tty)
		tty="/dev/ttyACM0";

	cc = cc_usb_open(tty);

	if (!cc)
		exit(1);

	/* And dump the resulting bits to the device
	 */

	while ((c = fread(buf, 1, sizeof(buf), file)) > 0)
		cc_usb_write(cc, buf, c);

	done(cc, 0);
}
